File: apps/ios/Tests/TenantKitTests/TestSupport.swift
import Foundation
import XCTest

#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

@testable import TenantKit

/// Records outgoing requests and replays canned responses.
final class StubTransport: HTTPTransport, @unchecked Sendable {
	struct Recorded {
		let request: URLRequest
		let body: [String: Any]
		let url: URL
	}

	private let lock = NSLock()
	private var responses: [(status: Int, data: Data, delayMs: Int)]
	private(set) var recorded: [Recorded] = []

	init(responses: [(status: Int, body: Any)] = []) {
		self.responses = responses.map { response in
			let data = (try? JSONSerialization.data(withJSONObject: response.body)) ?? Data()
			return (response.status, data, 0)
		}
	}

	init(rawResponses: [(status: Int, data: Data)]) {
		self.responses = rawResponses.map { ($0.status, $0.data, 0) }
	}

	/// Like `rawResponses`, but each reply can be delayed — used to pin down
	/// interleavings such as a refresh that lands after a sign-out.
	init(delayedResponses: [(status: Int, data: Data, delayMs: Int)]) {
		self.responses = delayedResponses.map { ($0.status, $0.data, $0.delayMs) }
	}

	func send(_ request: URLRequest) async throws -> (Data, HTTPURLResponse) {
		lock.lock()
		let response = responses.isEmpty ? (200, Data("{}".utf8), 0) : responses.removeFirst()
		lock.unlock()

		if response.2 > 0 {
			// Swallow cancellation: a reply already in flight still arrives, which
			// is the case the session-level sign-out guard has to handle.
			try? await Task.sleep(nanoseconds: UInt64(response.2) * 1_000_000)
		}

		lock.lock()
		let body: [String: Any] = {
			guard let data = request.httpBody else { return [:] }
			return (try? JSONSerialization.jsonObject(with: data) as? [String: Any]) ?? [:]
		}()
		if let url = request.url {
			recorded.append(Recorded(request: request, body: body, url: url))
		}
		lock.unlock()

		let http = HTTPURLResponse(
			url: request.url ?? URL(string: "https://example.com")!,
			statusCode: response.0,
			httpVersion: "HTTP/1.1",
			headerFields: ["Content-Type": "application/json"]
		)!
		return (response.1, http)
	}

	var lastRequest: Recorded? {
		lock.lock()
		defer { lock.unlock() }
		return recorded.last
	}
}

extension StubTransport {
	static func json(_ object: Any, status: Int = 200) -> (status: Int, body: Any) {
		(status, object)
	}
}

/// Always fails, like an offline device.
struct FailingTransport: HTTPTransport {
	func send(_ request: URLRequest) async throws -> (Data, HTTPURLResponse) {
		throw URLError(.notConnectedToInternet)
	}
}

func makeConfiguration(
	slug: String? = "acme",
	host: String? = nil,
	origin: String? = nil,
	region: String = "us"
) -> TenantConfiguration {
	TenantConfiguration(
		appBaseURL: URL(string: "https://app.epic-startup.com")!,
		tenantAPIBaseURLUS: URL(string: "https://tenant-us.epic-startup.com")!,
		tenantAPIBaseURLKSA: URL(string: "https://tenant-ksa.epic-startup.com")!,
		siteAddress: SiteAddress(
			slug: slug,
			host: host,
			origin: origin.flatMap(URL.init(string:))
		),
		brandDomain: "epic-startup.com",
		dataRegion: region
	)
}

/// Unsigned JWT with the same claim shape tenant-api issues.
func makeAccessToken(
	customerId: String = "cust_1",
	orgId: String = "org_1",
	name: String = "Ada Lovelace",
	expiresIn: TimeInterval = 900
) -> String {
	let header = #"{"alg":"HS256","typ":"JWT"}"#
	let payload = """
	{"customerId":"\(customerId)","orgId":"\(orgId)","name":"\(name)","type":"access","exp":\(Int(Date().timeIntervalSince1970 + expiresIn))}
	"""
	func encode(_ value: String) -> String {
		let data = Data(value.utf8)
		return data.base64EncodedString()
			.replacingOccurrences(of: "+", with: "-")
			.replacingOccurrences(of: "/", with: "_")
			.replacingOccurrences(of: "=", with: "")
	}
	return "\(encode(header)).\(encode(payload)).signature"
}

File: apps/android/app/src/test/kotlin/com/epicstartup/tenant/core/TestSupport.kt
package com.epicstartup.tenant.core

import com.epicstartup.tenant.core.config.TenantConfiguration
import com.epicstartup.tenant.core.net.ApiRequest
import com.epicstartup.tenant.core.net.ApiResponse
import com.epicstartup.tenant.core.net.HttpTransport
import com.epicstartup.tenant.core.support.SiteAddress
import java.util.Base64
import java.util.concurrent.CopyOnWriteArrayList
import org.json.JSONObject

/** Records outgoing requests and replays canned responses. */
class StubTransport(
	private val handler: (ApiRequest) -> ApiResponse = { ApiResponse(200, ByteArray(0)) },
) : HttpTransport {
	private val recorded = CopyOnWriteArrayList<ApiRequest>()

	override fun send(request: ApiRequest): ApiResponse {
		recorded += request
		return handler(request)
	}

	val requests: List<ApiRequest> get() = recorded.toList()

	val lastRequest: ApiRequest? get() = recorded.lastOrNull()
}

fun jsonResponse(status: Int, body: String): ApiResponse =
	ApiResponse(statusCode = status, body = body.toByteArray(Charsets.UTF_8))

fun jsonBody(vararg pairs: Pair<String, Any?>): String = JSONObject().apply {
	pairs.forEach { (key, value) -> if (value != null) put(key, value) }
}.toString()

fun requestBodyJson(request: ApiRequest): JSONObject =
	JSONObject(String(request.body ?: ByteArray(0), Charsets.UTF_8))

fun makeConfiguration(
	slug: String? = "acme",
	host: String? = null,
	origin: String? = "https://acme.epic-startup.com",
	dataRegion: String = "us",
	appBaseUrl: String = "https://app.epic-startup.com",
	tenantApiUs: String = "https://tenant-us.epic-startup.com",
	tenantApiKsa: String? = "https://tenant-ksa.epic-startup.com",
): TenantConfiguration = TenantConfiguration(
	appBaseUrl = appBaseUrl,
	tenantApiBaseUrlUs = tenantApiUs,
	tenantApiBaseUrlKsa = tenantApiKsa,
	siteAddress = SiteAddress(slug = slug, host = host, origin = origin),
	dataRegion = dataRegion,
)

fun base64Url(value: String): String =
	Base64.getUrlEncoder().withoutPadding().encodeToString(value.toByteArray(Charsets.UTF_8))

/** Builds a (deliberately unsigned) JWT shaped like tenant-api's access token. */
fun makeAccessToken(
	customerId: String = "cus_1",
	orgId: String = "org_1",
	name: String? = "Jane Doe",
	expiresAtSeconds: Long = System.currentTimeMillis() / 1000 + 900,
): String {
	val payload = JSONObject().apply {
		put("customerId", customerId)
		put("orgId", orgId)
		if (name != null) put("name", name)
		put("exp", expiresAtSeconds)
	}
	return "${base64Url("""{"alg":"HS256","typ":"JWT"}""")}.${base64Url(payload.toString())}.signature"
}
